Contribute `skyAppLinkExternal`

Blackbaud-MitchellThomas opened the pull request:
We need to navigate to an other SPA but maintain service and environment context. So, we need to sidestep the Angular router but still get the same behavior as `skyAppLink` to maintain the runtime context for the SPA.

- First commit for review on overall approach before writing tests

Blackbaud-PaulCrowder commented on runtime/directives/sky-app-link-external.directive.ts:
```diff
@@ -0,0 +1,36 @@
+})
+export class SkyAppLinkExternalDirective implements OnInit {
+  @Input() private skyAppLinkExternal: string;
+  private externalSPAUrl: string;
```
To remain consistent with terminology used across the rest of the app, this should be named `externalAppUrl` or simply `externalUrl`.

Blackbaud-PaulCrowder commented on runtime/directives/sky-app-link-external.directive.ts:
```diff
@@ -0,0 +1,36 @@
+  selector: '[skyAppLinkExternal]'
+})
+export class SkyAppLinkExternalDirective implements OnInit {
+  @Input() private skyAppLinkExternal: string;
```
This should be refactored to have a `get`ter and a `set`ter so that the `externalUrl` property can be updated whenever the value changes.  Right now you're only setting the URL when the directive is initialized.

Blackbaud-PaulCrowder commented on runtime/directives/sky-app-link-external.directive.ts:
```diff
@@ -0,0 +1,36 @@
+
+  @HostListener('click', ['$event'])
+  public onClick(event: MouseEvent) {
+    this.window.nativeWindow.location.href = this.externalSPAUrl;
```
There's no need to handle the element's `click` event; simply update the element's `href` property which you can do using a `@HostBinding()` property.  See Angular's implementation of `routerLink` for an example.

https://github.com/angular/angular/blob/5.2.7/packages/router/src/directives/router_link.ts#L182
https://github.com/angular/angular/blob/5.2.7/packages/router/src/directives/router_link.ts#L233

Blackbaud-PaulCrowder commented on runtime/directives/sky-app-link-external.directive.ts:
```diff
@@ -0,0 +1,36 @@
+import { HttpParams } from '@angular/common/http';
+
+@Directive({
+  selector: '[skyAppLinkExternal]'
```
Make the selector more specific to only allow anchor tags (i.e. `a[skyAppLinkExternal]`) until there's a compelling use case for non-anchor elements.

Blackbaud-PaulCrowder commented on runtime/directives/sky-app-link-external.directive.ts:
```diff
@@ -0,0 +1,36 @@
+
+  public ngOnInit() {
+    const queryParams: any = this.skyAppConfig.runtime.params.getAll();
+    const urlParts = this.skyAppLinkExternal.split('?');
```
This assumes the `skyAppLinkExternal` value is a string, but Angular's `routerLink` (and therefore `skyAppLink`) also allow arrays so that URLs can be built up in parts that are dynamic for consumer convenience.  For instance, this is valid where `part2` is a public property of the component:

`<a [skyAppLink]="['part-1', part2, 'part-3']">Dynamic link</a>`

Blackbaud-MitchellThomas commented:
@Blackbaud-PaulCrowder, I addressed your feedback and have written tests against the new approach, which leverages most of the url logic behind `RouterLinkWithHref`

Blackbaud-BobbyEarl commented:
Hey @Blackbaud-MitchellThomas this looks good to me.  Looks like there's a conflict with master now.  Can you get that addressed?  We can then get it approved and merged in.

Blackbaud-BobbyEarl commented:
@Blackbaud-PaulCrowder Can you confirm the current state meets your "changes requested?"
